Alternative gen for .msg and conversion headers

v0-e opened the pull request:
Adds new code generators for ROS `.msg` and asn1c structs to ROS structs conversion headers, as discussed in #13.
Implements two compiler backends for [librasn's compiler](https://github.com/librasn/compiler). As I used a bit of the code written by @6d7a for the Rust backend, I'm adding him as a co-author.

Code generation is pretty much as asn1c:
- Compile the two new code generators using Docker using `utils/codegen/docker/rgen.Dockerfile`;
- Use `asn1ToRosMsg-rgen.py` to generate the ROS messages;
- Use `asn1ToRConversionHeader-rgen.py` to generate the conversion headers;

Specify a `-m`/`--message` (e.g., `cam`, `denm`) and the newly built container using `-di` for the above Python scripts.

Let me know what you think :+1: 
Any issues please report them. If you prefer other naming for variables/functions/files please also let me know.
I'll keep this PR as a draft for now.

v0-e commented:
Added/renamed constraint-related fields to better comply with the expected naming used in `etsi_its_msgs_utils`.

v0-e commented:
Tested full CAM conversion pipeline with the provided RViz plugin. Vehicles are displayed correctly. 👍
Tested real-time using simulated OBUs sending CAMs: OBUs -> RSU -> UPER to ROS Conversion node -> RViz

lreiher commented:
Awesome to hear that the testing is looking good! It's very cool to see that you made such large progress in that short amount of time.

I've already spent a couple of minutes to try to regenerate CAM myself (working fine). Before going into further testing and also before looking at any other details, I'd like to ask you to do the following:
- If possible, could you please add [VS Code `launch.json`](https://github.com/ika-rwth-aachen/etsi_its_messages/blob/main/.vscode/launch.json) debugging settings? That would also simplify testing for us.
- Could you please re-generate CAM and DENM (c, msgs, conversion headers) using your approach and then commit those changes? I'd like to have a look at the diff between the current code and the one generated by your implementation.

Thanks again!

v0-e commented:
Hey @lreiher,
- I don't really use VS Code, though added run options for the new generation. Didn't test it but should work I think. Let me know. Interfacing the `.py` generator helpers should be the same now (I've removed the `-m` option mentioned in the OP);
- Added the new `.msg`s and conversion headers. The formatting still needs some more work, and the generated files are currently missing the license. Let me know what you think. Regarding the c generation, I didn't pushed anything yet. I'm using the asn1c's fork of [mouse07410](https://github.com/mouse07410/asn1c) though should be equivalent to fillabs' (atleast for the CAMs/DENMs).

lreiher commented:
Hi, your PR hasn't been forgotten yet. ;) Sorry for the delay, but we have now had the time to have a closer look and will add a few more review comments on Monday.

We are really looking forward to merge this (as an alternative for now) as soon as possible, with the potential of also completely switching to your approach in the near future.

Have a great weekend!

lreiher commented on utils/codegen/asn1ToRosMsg-rgen.py:
```diff
@@ -0,0 +1,89 @@
+    # create output directory
+    os.makedirs(args.output_dir, exist_ok=True)
+
+    # create temporary directories for running asn1c in docker container
```
There's some copy-paste comments here in this file that still need to be adjusted.

v0-e replied:
Fixed.

lreiher commented on utils/codegen/asn1ToConversionHeader-rgen.py:
```diff
@@ -0,0 +1,90 @@
+    parser.add_argument("files", type=str, nargs="+", help="ASN1 files")
+    parser.add_argument("-o", "--output-dir", type=str, required=True, help="output directory")
+    parser.add_argument("-td", "--temp-dir", type=str, default=None, help="temporary directory for mounting files to container; uses tempfile by default")
+    parser.add_argument("-t", "--message", type=str, required=True, help="message name")
```
```suggestion
    parser.add_argument("-m", "--message", type=str, required=True, help="message name")
```

v0-e replied:
Changed.

lreiher commented on utils/codegen/asn1ToConversionHeader-rgen.py:
```diff
@@ -0,0 +1,90 @@
+    # create output directory
+    os.makedirs(args.output_dir, exist_ok=True)
+
+    # create temporary directories for running asn1c in docker container
```
There's some copy-paste comments here in this file that still need to be adjusted.

v0-e replied:
Fixed.

lreiher commented on utils/codegen/docker/rgen.Dockerfile:
```diff
@@ -0,0 +1,54 @@
+# docker build -t ghcr.io/ika-rwth-aachen/etsi_its_messages:rgen -f rgen.Dockerfile .
```
We should poush this image to the repo's container registry.

We can leave this thread open until all other are resolved and we are ready to merge.

lreiher reviewed (commented):
Hi, I have now left a few review comments. :)

Again, thank you very much, we were able to directly run the generation using the VS Code configuration.

Our main request right now would be to try to match the existing generated CAM/DENM code files as closely as possible, see review comments.

We would then be more than happy to merge this as soon as possible. In the beginning, we would probably still keep both alternatives for code generation. As we then see new messages (e.g., SPAT, CPM, ...) being added, we are also open to completely replacing the Python-based approach with your Rust-based one.

v0-e commented on etsi_its_msgs/etsi_its_cam_msgs/msg/CamParameters.msg:
```diff
@@ -1,46 +1,20 @@
-# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
-# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
-# SOFTWARE.
-# ==============================================================================
```
Added.

v0-e commented on etsi_its_msgs/etsi_its_cam_msgs/msg/CamParameters.msg:
```diff
@@ -1,46 +1,20 @@
-# SOFTWARE.
-# ==============================================================================
-
-# --- Auto-generated by asn1ToRosMsg.py ----------------------------------------
```
Added.

v0-e commented on etsi_its_conversion/etsi_its_cam_conversion/include/etsi_its_cam_conversion/convertCamParameters.h:
```diff
@@ -43,39 +22,31 @@ namespace cam_msgs = etsi_its_cam_msgs::msg;
   toStruct_HighFrequencyContainer(in.high_frequency_container, out.highFrequencyContainer);
   if (in.low_frequency_container_is_present) {
-    LowFrequencyContainer_t low_frequency_container;
-    toStruct_LowFrequencyContainer(in.low_frequency_container, low_frequency_container);
-    out.lowFrequencyContainer = new LowFrequencyContainer_t(low_frequency_container);
+    out.lowFrequencyContainer = (LowFrequencyContainer_t*) calloc(1, sizeof(LowFrequencyContainer_t));
+    toStruct_LowFrequencyContainer(in.low_frequency_container, *out.lowFrequencyContainer);
```
I'm guessing there are memory leaks for messages requiring dynamic allocation (messages with optional fields, sequences ofs) in the current conversion node code, as there are no matching `delete`s for the `new`s. 
`asn1c` provides methods to [deep free structs](https://github.com/mouse07410/asn1c/blob/5fa129cf3161da9ff2d67341e6277e818c0f5d6a/skeletons/constr_TYPE.h#L85), however these methods use [by default](https://github.com/mouse07410/asn1c/blob/5fa129cf3161da9ff2d67341e6277e818c0f5d6a/skeletons/asn_internal.h#L40) `free()`.
The new `calloc` here should allow us to the cleanup methods like `ASN_STRUCT_FREE` or `ASN_STRUCT_FREE_CONTENTS_ONLY`.

lreiher replied:
Good catch, unfortunately, we are already aware of that memory leak. We were considering wrapping the structs in yet another type providing a destructor for cleanup.

However, we didn't know about the existing methods for deep freeing structs yet, thanks! Does that mean that we can simply call those with a `CAM_t`? If yes, we should also integrate that into the conversion node to avoid the memory from leaking there.

v0-e replied:
Yeah, it is easy to integrate. I can do it either on this PR or create another whenever this one is accepted.

v0-e commented on etsi_its_conversion/etsi_its_cam_conversion/include/etsi_its_cam_conversion/convertAccelerationConfidence.h:
```diff
@@ -1,31 +1,10 @@
 
 #pragma once
 
+#include <stdexcept>
```
It is required in some conversions that can `throw`, e.g., [here](https://github.com/ika-rwth-aachen/etsi_its_messages/blob/428f3bb4afc430385fac7d90738e2cb9cdbc8339/etsi_its_conversion/etsi_its_cam_conversion/include/etsi_its_cam_conversion/convertEventHistory.h#L67).
I can try to only add the include to the relevant headers.

v0-e replied:
Fixed.

v0-e commented on utils/codegen/asn1ToRosMsg-rgen.py:
Done.

lreiher commented on utils/codegen/docker/rgen/README.md:
We also need to verify that all README commands are still using the correct paths.

v0-e replied:
I've changed the README commands to point to the Python gen.

v0-e commented on etsi_its_msgs/etsi_its_cam_msgs/msg/CamParameters.msg:
```diff
@@ -1,46 +1,20 @@
-# 	...
-# }
-# ------------------------------------------------------------------------------
+## SEQUENCE CamParameters .extensible
```
With the Rust compiler it is currently not possible. I've added the use of your parser instead.

v0-e commented on etsi_its_msgs/etsi_its_cam_msgs/msg/CamParameters.msg:
```diff
@@ -1,46 +1,20 @@
+
+
+
+
```
Should be better match now overall.

v0-e commented on etsi_its_msgs/etsi_its_cam_msgs/msg/CamParameters.msg:
```diff
@@ -1,46 +1,20 @@
 
-SpecialVehicleContainer special_vehicle_container
 bool special_vehicle_container_is_present
+SpecialVehicleContainer special_vehicle_container
```
The order should be about the same now.

v0-e commented on etsi_its_conversion/etsi_its_cam_conversion/include/etsi_its_cam_conversion/convertCamParameters.h:
Gotcha.

v0-e commented on utils/codegen/docker/rgen/README.md:
I should also do some code cleanup whenever you agree this is ready to merge. :+1:

v0-e commented on etsi_its_msgs/etsi_its_cam_msgs/msg/CamParameters.msg:
Should be a better match overall now.
Some more noticeable differences:
1. Some fields/vars are converted differently like [this](https://github.com/v0-e/etsi_its_messages/blob/d8034bd0bfa11a192fc9b7510aa2b3962e15b067/etsi_its_msgs/etsi_its_cam_msgs/msg/PositioningSolutionType.msg#L36). Special cases like this one are currently [converted "manually"](https://github.com/v0-e/etsi_its_messages/blob/d8034bd0bfa11a192fc9b7510aa2b3962e15b067/etsi_its_msgs/etsi_its_cam_msgs/msg/PositioningSolutionType.msg#L36) in the Python gen.
2. The conversion headers have newlines in different places inside the conversion functions. This is a bit more difficult to match, as I used kind of a single template for both primitives and non-primitives conversion. The final output of the Rust gen should also be "pretty" enough.
3. There are some integer types which are converted to `int64` in the Rust gen, instead of a shorter integer like in [here](https://github.com/v0-e/etsi_its_messages/blob/rust-based-codegen/etsi_its_msgs/etsi_its_cam_msgs/msg/PathDeltaTime.msg#L32). The Rust compiler converts these to their larger representation as they are extensible, in order to provide some compatibility with other future definitions of the same type in case they are extended in range.

v0-e commented:
Hi @lreiher,
Thanks for the comments :)
I think I've tackled most of your concerns. Let me know what you think :+1:
